sway-core/src/decl_engine/wrapper.rs

```diff
@@ -56,6 +53,52 @@ impl PartialEqWithEngines for DeclWrapper {
     }
 }
 
+impl HashWithEngines for DeclWrapper {
+    fn hash<H: Hasher>(&self, state: &mut H, engines: Engines<'_>) {
+        match self {
+            DeclWrapper::Unknown => {
+                state.write_u8(self.discriminant_value());
+            }
+            DeclWrapper::Function(decl) => {
+                state.write_u8(self.discriminant_value());
+                decl.hash(state, engines);
+            }
+            DeclWrapper::Trait(decl) => {
+                state.write_u8(self.discriminant_value());
+                decl.hash(state, engines);
+            }
+            DeclWrapper::TraitFn(decl) => {
+                state.write_u8(self.discriminant_value());
+                decl.hash(state, engines);
+            }
+            DeclWrapper::ImplTrait(decl) => {
+                state.write_u8(self.discriminant_value());
+                decl.hash(state, engines);
+            }
```

Comment:
This might be subjective, but I'd whip up a little macro for most of these (and the others in other types in this PR).